### PR TITLE
hygiene: Implement transparent marks and use them for call-site hygiene in proc-macros

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes https://github.com/rust-lang/rust/issues/50050

### CODE REVIEW

**Comment by petrochenkov** [comment]
This needs a check-only crater run.
ping @rust-lang/infra

**Comment by petrochenkov** [comment]
CI failure fixed.
@bors try

**Comment by alexreg** [comment]
@petrochenkov Looks like this will be relevant to my PR https://github.com/rust-lang/rust/pull/47992? (I still need to write up the RFC for this, I know.)

**Comment by petrochenkov** [comment]
@alexreg 
Yes, `#[rustc_transparent_macro]` introduced in this PR for testing purposes is exactly the decl macro hygiene opt-out, but applied to the whole right-hand side of the macro.

**Comment by alexreg** [comment]
@Petrochenkov, But I mean, I should probably adapt my PR to use transparent marks on syntax contexts now, for hygiene-escaped idents, right? (Rather than no marks like now.)

**Comment by petrochenkov** [comment]
@alexreg 
>I should probably adapt my PR to use transparent marks on syntax contexts now, for hygiene-escaped idents, right? (Rather than no marks like now.)

Yes.

**Comment by pietroalbini** [comment]
Check-only run started on `crater-1`. It should finish in ~2 days.

**Comment by pietroalbini** [comment]
Hi @petrochenkov  (crater requester), @oli-obk  (PR reviewer)! Crater results are at: http://cargobomb-reports.s3.amazonaws.com/pr-51762/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/config.toml). If you see any spurious failures not on the list, please make a PR against that file.

(interested observers: Crater is a tool for testing the impact of changes on the crates.io ecosystem. You can find out more at the [repo](https://github.com/rust-lang-nursery/crater/) if you're curious)

**Comment by oli-obk** [comment]
The uom crate failed (http://cargobomb-reports.s3.amazonaws.com/pr-51762/d20562b7ebf6ba2344a2b578710440deee3a92bb/reg/uom-0.19.0/log.txt) with a lot of macro related errors

**Comment by petrochenkov** [comment]
Nice, only one regression!
I'll look what happens with `uom`.

**Comment by petrochenkov** [comment]
Updated with a fix for `uom`.

File: src/libsyntax_pos/symbol.rs [hunk 1]
```diff
@@ -68,6 +68,15 @@ impl Ident {
         Ident::new(self.name, self.span.modern())
     }
 
+    // "Normalize" ident for use in comparisons using "local variable hygiene".
```
Comment (oli-obk) [inline src/libsyntax_pos/symbol.rs]: This can be a doc comment

File: src/test/ui/hygiene/auxiliary/transparent-basic.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(decl_macro, rustc_attrs)]
+
+#[rustc_transparent_macro]
```
Comment (oli-obk) [inline src/test/ui/hygiene/auxiliary/transparent-basic.rs]: Is this attribute only meant for testing?

If so, the rustc-guide should probably get an entry about it

File: src/test/ui/hygiene/transparent-basic.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+fn legacy_interaction1() {
+    legacy!();
+}
+
+struct S;
+
+fn check_dollar_crate() {
+    // `$crate::S` inside the macro resolves to `S` from this crate.
```
Comment (oli-obk) [inline src/test/ui/hygiene/transparent-basic.rs]: That is so confusing. Do we have any canonical macro documentation? I feel very much out of my depths here.

File: src/libproc_macro/lib.rs [hunk 4]
```diff
@@ -1430,24 +1435,39 @@ pub mod __internal {
 
         CURRENT_SESS.with(|p| {
             let _reset = Reset { prev: p.get() };
-            p.set((cx.parse_sess, cx.current_expansion.mark));
+
+            // No way to determine def location for a proc macro rigth now, so use call location.
```
Comment (oli-obk) [inline src/libproc_macro/lib.rs]: typo: "rigth"

File: src/libproc_macro/lib.rs [hunk 5]
```diff
@@ -1430,24 +1435,39 @@ pub mod __internal {
...
+            let location = cx.current_expansion.mark.expn_info().unwrap().call_site;
+            // Opaque mark was already created by expansion, now create its transparent twin.
```
Comment (oli-obk) [inline src/libproc_macro/lib.rs]: Can you add some explanation here as to the "why" we need the transparent mark?

File: src/librustc_resolve/lib.rs [hunk 6]
```diff
@@ -1989,7 +1989,31 @@ impl<'a> Resolver<'a> {
             // When resolving `$crate` from a `macro_rules!` invoked in a `macro`,
             // we don't want to pretend that the `macro_rules!` definition is in the `macro`
             // as described in `SyntaxContext::apply_mark`, so we ignore prepended modern marks.
-            ctxt.marks().into_iter().rev().find(|m| m.transparency() != Transparency::Transparent)
+            // FIXME: This is only a guess and it doesn't work correctly for `macro_rules!`
```
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: Can we just forbid `macro` to produce `macro_rules` and `macro_rules` to produce `macro`? Or is that something we'll want to support

File: src/librustc_resolve/lib.rs [hunk 7]
```diff
@@ -1989,7 +1989,31 @@ impl<'a> Resolver<'a> {
...
+            while let Some(&mark) = iter.peek() {
+                if mark.transparency() == Transparency::Opaque {
+                    result = Some(mark);
+                    iter.next();
+                } else {
+                    break;
+                }
+            }
+            // Then find the last legacy mark from the end if it exists.
+            while let Some(&mark) = iter.peek() {
```
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: this can just be a for loop over the iterator

**Review by oli-obk — CHANGES_REQUESTED** [review]
After reading this for the third time I think I get now what's going on.

I'd really love it if we could start building some content in https://rust-lang-nursery.github.io/rustc-guide/macro-expansion.html#hygiene

You don't have to do all of it, just the logic that is applied in this PR and some links into the code so it's easier to navigate.

File: src/test/ui/hygiene/auxiliary/transparent-basic.rs (hunk 2 above)
Comment (petrochenkov) [reply src/test/ui/hygiene/auxiliary/transparent-basic.rs]: Yes, for testing.
It's very inconvenient to only be able to test transparent marks on stage2 with proc macros.
It's going to be replaced with more fine-grained hygiene opt-out (https://github.com/rust-lang/rust/pull/47992) eventually.

File: src/libproc_macro/lib.rs (hunk 5 above)
Comment (petrochenkov) [reply src/libproc_macro/lib.rs]: Do you mean why it's transparent, or why we need a separate mark at all, or why we can't just use the call-site span literally?

File: src/librustc_resolve/lib.rs (hunk 6 above)
Comment (petrochenkov) [reply src/librustc_resolve/lib.rs]: I still hope to fix this by replacing the context transplantation trick (see below) with some less ambiguous representation.

File: src/test/ui/hygiene/transparent-basic.rs (hunk 3 above)
Comment (petrochenkov) [reply src/test/ui/hygiene/transparent-basic.rs]: This means transparent macros generate *everything* at call-site without exception, including `$crate`.

>Do we have any canonical macro documentation?

Unfortunately no, but I'm going to write something soon™!

**Comment by petrochenkov** [comment]
>I'd really love it if we could start building some content in https://rust-lang-nursery.github.io/rustc-guide/macro-expansion.html#hygiene

Yes, I want to write some docs about the implementation, and an RFC for macros 2.0 with detailed specification eventually, but hygiene is in kind of transitional state right now.
I want to fix https://github.com/rust-lang/rust/issues/50504 first, then hopefully eliminate the context transplantation trick (see below), then document the result in rustc guide.

File: src/libproc_macro/lib.rs (hunk 5 above)
Comment (oli-obk) [reply src/libproc_macro/lib.rs]: The latter, why do we need to anything special here.

**Comment by oli-obk** [comment]
Ok... seems fine to me then.

r=me with the nits.

**Comment by petrochenkov** [comment]
### The "context transplantation trick"

Preface: syntactic context
```
macro x() {
    macro y()  {
        macro z() {
            my_token
        }
    }
}

...
// produces `my_token`, `my_token` has context `root -> x -> y -> z`
// i.e. context corresponds to a sequence of nested macro *definitions*
z!();
```

When we have this situation:
```rust
macro_rules! less_opaque_macro_call { // y
    () => { my_token }
}

macro more_opaque_macro_def() { // x
    less_opaque_macro_call!();
}

...
// would produce `my_token`, `my_token` would have context `root -> y`, BUT...
less_opaque_macro_call!();
```
we transform the code above into something like this:
```rust
macro more_opaque_macro_def() { // x
    macro_rules! less_opaque_macro_call { // y
        () => { my_token }
    }
    less_opaque_macro_call!();
}

// now produces `my_token`, `my_token` has context `root -> x -> y`
less_opaque_macro_call!();
```
to achieve the observable effect described in https://github.com/rust-lang/rust/pull/46551, i.e. "modern" hygienic macros "containing and restraining" non-hygienic macros called inside them.
However, the resulting representation `root -> x -> y` is now ambiguous because it may represent both 1) `macro_rules` actually defined inside `macro` and 2) `macro_rules` transplanted inside of `macro`.
Due to this ambiguity we have issues with `$crate` and I suspect other issues in complex nesting cases.

I want to find some other representation for hygiene data so that the hygiene algorithm could work directly on def-site and call-site hierarchies without doing the transplantation.

**Pull request merged**